hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1300,6 +1300,79 @@ private class Emit(
             res.isNull))
 
         EmitTriplet(setup, m, res.invoke[Double]("doubleValue"))
+      case x@MakeNDArray(dataIR, shapeIR, rowMajorIR) =>
+        val dataContainer = dataIR.pType.asInstanceOf[PArray]
+        val shapePType = shapeIR.pType.asInstanceOf[PTuple]
+        val nDims = shapePType.size
+
+        val datat = emit(dataIR)
+        val shapet = emit(shapeIR)
+        val rowMajort = emit(rowMajorIR)
+
+        val t = x.pType.asInstanceOf[PNDArray]
+        val srvb = new StagedRegionValueBuilder(mb, t.representation)
+
+        val setup = Code(
+          shapet.setup,
+          datat.setup,
+          rowMajort.setup)
+        val value = coerce[Long](Code(
+          srvb.start(),
+          srvb.addInt(0),
+          srvb.advance(),
+          srvb.addInt(0),
+          srvb.advance(),
+          shapet.m.mux(
+            Code._fatal("Missing shape"),
+            Code(
+              srvb.addBaseStruct(t.representation.fieldType("shape").asInstanceOf[PBaseStruct], { srvb: StagedRegionValueBuilder =>
+                Code(
+                  srvb.start(),
+                  {
+                    var index = 0
+                    var shapeCopyingCode = Code._empty[Unit]
+                    while (index < nDims) {
```

Comment:
One way to make this a little cleaner:
```scala
Code(
  srvb.start(),
  Code(Array.tabulate(nDims) { index => Code(
    shapePType.isFieldMissing(shapet.value[Long], index).mux[Unit](
	Code._fatal(s"shape missing at index $index"),
    Code(
	  srvb.addLong(region.loadLong(shapePType.loadField(coerce[Long](shapet.v), index))), 
	  srvb.advance())
  )}: _*))
```
Might be nice to add something like
```scala
foreach[A](it: Seq[A])(f: A => Code[_]): Code[_] = Code(it.map(f): _*)
```
to `Code`. Then the above would be a bit more readable:
```scala
Code(
  srvb.start(),
  Code.foreach(Range(0, nDims)) { index => Code(
    ...
  )}))
```

Reply:
I like this, look much better.